### PR TITLE
MFEM export

### PR DESCRIPTION
**Pull request by jzwar**
# MFEM export
- New function to export lists of arbitrary spline types into MFEM-format. Related to issue #31 part 1
- Updated JSON export and import

## MFEM export
First step on the way to creating a bullet proof mfem export. See #31 for more details.
Currently, only structured spline-patches are supported. If this is not the case, an exception is thrown from the bezman library. It is currently not catched or passed, which is something that needs to be implemented still.

Required work on this PR:
- extend unit tests for 3D meshes
- check with respect to reference solution (mfem export?)
- JSON export is only a draft up until now (still required?)

### CODE REVIEW

File: splinepy/io/mfem.py [hunk 1]
```diff
@@ -269,14 +269,176 @@ def flatten(l):
...
+      file name
+    bezier_list: list
+      list of bezier spline objects
+
+    Returns
+    -------
+    None
+    """
+    from splinepy._spline import Spline
+
```
Comment (j042) [inline splinepy/io/mfem.py:296]:
```suggestion
def export_cartesian(fname, spline_list, tolerance=None):
    """
    Export list of bezier splines in mfem export

    Parameters
    ----------
    fname: string
      file name
    bezier_list: list
      list of bezier spline objects

    Returns
    -------
    None
    """
    from splinepy._spline import Spline

    if tolerance is None:
        tolerance = 1e-5  #  or maybe settings.TOLERANCE 
```

**Comment by j042** [comment]
Good Stuff!

**Comment by jzwar** [comment]
@j042 This is my suggestion for mfem export functionality. I am open for further improvements

File: cpp/splinepy/py/spline_exporter.hpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    int* py_boundaries_ptr = static_cast<int*>(py_boundaries.request().ptr);
+    for (std::size_t i_boundary{}; i_boundary < boundaries.size();
+         i_boundary++) {
+      for (std::size_t i_id{}; i_id < 2ul; i_id++) {
+        py_boundaries_ptr[i_boundary * 2ul + i_id] =
+            static_cast<int>(boundaries[i_boundary][i_id]);
+      }
+    }
+
+    return std::make_tuple(py_connectivity, py_vertex_ids, py_edges,
```
Comment (j042) [inline cpp/splinepy/py/spline_exporter.hpp:141]:
```suggestion
    return py::make_tuple(py_connectivity, py_vertex_ids, py_edges,
```

**Review by j042 — COMMENTED** [review]
Thank you! This is an awesome feature! I will take further look at it sometime this week

File: cpp/splinepy/py/spline_exporter.hpp [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+namespace py = pybind11;
+
+/**
+ * @brief Retrieve information regarding export
+ *
+ * @param coordinates
+ *
+ */
+std::tuple<py::array_t<int>,  // connectivity
```
[the remark this answers is not in the document]
Comment (j042) [reply cpp/splinepy/py/spline_exporter.hpp:23]: verzeihung, I think the correct format is just `py::tuple`

File: splinepy/io/mfem.py [hunk 4]
```diff
@@ -269,14 +269,217 @@ def flatten(l):
...
+                np.isin(
+                    boundaries,
+                    boundary_vertex_ids[is_on_boundary]
+                ), axis=1) == n_vertex_per_boundary
+            boundary_ids[set_boundary] = i_bound + 2
+
+    # Check if algorithm was successfull
+    if not success:
+        # todo : only connectivity stores data, see issue #33
+        print(connectivity)
```
Comment (j042) [inline splinepy/io/mfem.py:404]: either debug logging or rm

File: tests/test_mfem_export.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
+import splinepy
+import numpy as np
+try:
+    from . import common as c
+except:
+    import common as c
+
+
+class ProximityTest(c.unittest.TestCase):
```
Comment (j042) [inline tests/test_mfem_export.py:9]:
```suggestion
class MFEMExportTest(c.unittest.TestCase):
```

File: tests/test_mfem_export.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        bsp_el2.elevate_degree(0)
+
+        bsp_el2.insert_knots(1, [0.5])
+        nur_el3.insert_knots(1, [0.5])
+
+        # Test Output
+        # @todo
+        splinepy.io.mfem.export_cartesian(
+            "test_mfem_r0.mesh", [bez_el0, bsp_el2, nur_el3, rbz_el1])
+
```
Comment (j042) [inline tests/test_mfem_export.py:68]: could you add a mfem exported & accepted file as a string for a butal and direct string comparison for now?

**Comment by j042** [comment]
Until we check all the boxes at #31, I can be satisfied with current status. minor fixes and let's merge :D

**Comment by j042** [comment]
Could you please add missing 10Px10D rational bezier splines (for non `--minimal` build)?

**Review by j042 — APPROVED** [review]
merging

**Pull request merged**